### ISSUE TITLE
envoy exited: signal: segmentation fault (core dumped)

### ISSUE THREAD

**Issue by dmitrif** [pomerium/pomerium#827]
## What happened?
Running latest pomerium all-in-one docker image. Worked for over a week, suddenly started crashing.

## What did you expect to happen?
Stable run.

## How'd it happen?

On launch crashes with an envoy segmentation fault.

## What's your environment like?

- Docker w/ Nomad, Ubuntu 16.04.6 LTS
- Image id: 2b7a439d73ee

## What's your config.yaml?

```
autocert: false

authenticate_service_url: https://auth.internal.xxx.com
shared_secret: xxx
cookie_secret: xxx

idp_provider: "google"
idp_provider_url: "https://accounts.google.com"
idp_client_id: xxx
idp_client_secret: xxx
idp_service_account: xxx

certificates:
- cert: "/secrets/certs/auth.internal.xxx.com/cert.pem"
  key: "/secrets/certs/auth.internal.xxx.com/key.pem"
- cert: "/secrets/certs/xxx.internal.xxx.com/cert.pem"
  key: "/secrets/certs/xxx.internal.xxx.com/key.pem"
```

## What did you see in the logs?

```logs
{"level":"debug","service":"envoy","name":"config","time":"2020-06-05T00:49:56Z","message":"Create listen socket for listener grpc-ingress on address [::]:5443"}
{"level":"debug","service":"envoy","name":"config","time":"2020-06-05T00:49:56Z","message":"Set listener grpc-ingress socket factory local address to [::]:5443"}
{"level":"debug","service":"envoy","name":"config","time":"2020-06-05T00:49:56Z","message":"add active listener: name=grpc-ingress, hash=14473341851986078457, address=[::]:5443"}
{"level":"info","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"\"lds: add/update listener \\'grpc-ingress\\'\""}
{"level":"debug","service":"envoy","name":"init","time":"2020-06-05T00:49:56Z","message":"target LDS initialized, notifying init manager Server"}
{"level":"debug","service":"envoy","name":"init","time":"2020-06-05T00:49:56Z","message":"init manager Server initialized, notifying RunHelper"}
{"level":"info","service":"envoy","name":"config","time":"2020-06-05T00:49:56Z","message":"all dependencies initialized. starting workers"}
{"level":"debug","service":"envoy","name":"config","time":"2020-06-05T00:49:56Z","message":"starting worker 0"}
{"level":"debug","service":"envoy","name":"config","time":"2020-06-05T00:49:56Z","message":"Resuming discovery requests for type.googleapis.com/envoy.api.v2.RouteConfiguration"}
{"level":"debug","service":"envoy","name":"config","time":"2020-06-05T00:49:56Z","message":"gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) https-ingress: Failed to load incomplete certificate from /root/.cache/pomerium/envoy/files/tls-crt-ef46db3751d8e999.pem, "}
{"level":"debug","service":"envoy","name":"envoy","time":"2020-06-05T00:49:56Z"}
{"level":"debug","service":"envoy","name":"main","time":"2020-06-05T00:49:56Z","message":"worker entering dispatch loop"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"adding TLS initial cluster policy-1573522986a9fe7"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"adding TLS initial cluster policy-184de50be6b1db32"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"adding TLS initial cluster policy-815f9fb27010879a"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"adding TLS initial cluster policy-a02cfe28d636c65b"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"adding TLS initial cluster policy-a0a490e2bb871854"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"adding TLS initial cluster pomerium-authz"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"adding TLS initial cluster pomerium-control-plane-grpc"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"adding TLS initial cluster pomerium-control-plane-http"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"membership update for TLS cluster pomerium-control-plane-grpc added 1 removed 0"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"updating TLS cluster pomerium-control-plane-http"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"membership update for TLS cluster pomerium-control-plane-http added 1 removed 0"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"updating TLS cluster pomerium-authz"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"membership update for TLS cluster pomerium-authz added 1 removed 0"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"updating TLS cluster policy-184de50be6b1db32"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"updating TLS cluster policy-815f9fb27010879a"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"updating TLS cluster policy-a0a490e2bb871854"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"updating TLS cluster policy-a02cfe28d636c65b"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"updating TLS cluster policy-1573522986a9fe7"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"membership update for TLS cluster policy-1573522986a9fe7 added 1 removed 0"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"membership update for TLS cluster policy-815f9fb27010879a added 2 removed 0"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"membership update for TLS cluster policy-a02cfe28d636c65b added 1 removed 0"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"membership update for TLS cluster policy-184de50be6b1db32 added 1 removed 0"}
{"level":"debug","service":"envoy","name":"upstream","time":"2020-06-05T00:49:56Z","message":"membership update for TLS cluster policy-a0a490e2bb871854 added 1 removed 0"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] Caught Segmentation fault, suspect faulting address 0x0"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:92] Envoy version: 3504d40f752eb5c20bc2883053547717bcb92fd8/1.14.1/clean-getenvoy-902f20f-envoy/RELEASE/BoringSSL"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #0: __restore_rt [0x7f02384b80e0]"}
{"level":"debug","service":"envoy","name":"grpc","time":"2020-06-05T00:49:56Z","message":"completionThread running"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #1: Envoy::Extensions::AccessLoggers::GrpcCommon::GrpcAccessLoggerCacheImpl::getOrCreateLogger() [0x56402a292fa7]"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #2: Envoy::Extensions::AccessLoggers::GrpcCommon::GrpcAccessLoggerCacheImpl::getOrCreateLogger() [0x56402a293372]"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #3: std::__1::__function::__func<>::operator()() [0x56402a28f038]"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #4: std::__1::__function::__func<>::operator()() [0x56402ac8b678]"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #5: std::__1::__function::__func<>::operator()() [0x56402ac8c8a8]"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #6: Envoy::Event::DispatcherImpl::runPostCallbacks() [0x56402ad04aa6]"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #7: Envoy::Event::DispatcherImpl::run() [0x56402ad0493c]"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #8: Envoy::Server::WorkerImpl::threadRoutine() [0x56402acfa9a4]"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #9: Envoy::Thread::ThreadImplPosix::ThreadImplPosix()::$_0::__invoke() [0x56402b202dd3]"}
{"level":"debug","service":"envoy","name":"backtrace","time":"2020-06-05T00:49:56Z","message":"[bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #10: start_thread [0x7f02384ae4a4]"}
{"level":"fatal","error":"envoy exited: signal: segmentation fault (core dumped)","time":"2020-06-05T00:49:56Z","message":"cmd/pomerium"}
```

## Additional Context
- Attempted to manually remove images / containers are restart fresh... Same error.

**Comment by desimone** [comment]
@dmitrif Thanks for making an issue! 

This looks possibly related to certificates. 

```
{"level":"debug","service":"envoy","name":"config","time":"2020-06-05T00:49:56Z","message":"gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) https-ingress: Failed to load incomplete certificate from /root/.cache/pomerium/envoy/files/tls-crt-ef46db3751d8e999.pem, "}
```

Can you validate that the cert referenced is valid? 

Either way, we should not be segfaulting on an error here.

**Comment by dmitrif** [comment]
Hi, @desimone ! 

I’ll take a look at the container, but I was under the impression that gprc doesn’t use ssl in the all in one solution. Does it cache the specified certs specified in the config? 

Thank you!

**Comment by dmitrif** [comment]
Hi! I could not inspect the Pomerium container since it bails out after the error. So I created a basic helloworld container with all the same secret injection / host volumes. 

The passed-in certs are present and are valid. So the cert Envoy is having trouble finding may be internal.

**Comment by dmitrif** [comment]
One more update, downgrading to v0.8.3 resolves the issue.

**Comment by desimone** [comment]
@dmitrif This should be fixed in v0.9.1 and later with https://github.com/pomerium/pomerium/pull/876

Closing for now, but please reopen if that's not the case.
